torch/_inductor/fx_passes/bucketing.py

```diff
@@ -277,523 +225,364 @@ def _rank_idx_dict(group_name: str) -> dict[int, int]:
+    new_ag_in = new_ag_out.narrow(0, ag_input_numel * local_rank, ag_input_numel)
+    foreach_copy_dsts = torch.split(new_ag_in, ins_split_sizes)
+    torch._foreach_copy_(foreach_copy_dsts, ag_ins_flattened)
+    wait_tensor = torch.ops.c10d_functional.wait_tensor(
+        torch.ops._c10d_functional.all_gather_into_tensor_out.default(
+            new_ag_in, group_size, group_name, out=new_ag_out
+        )
+    )
+    new_ag_out_reshaped = wait_tensor.reshape(group_size, -1)
+    outs = torch.split_with_sizes(
+        new_ag_out_reshaped,
+        ins_split_sizes,
+        dim=1,
+    )
+    outs_reshaped = [
+        o.reshape((shape[0] * group_size,) + shape[1:])
+        for o, shape in zip(outs, ins_sizes)
+    ]
+    return outs_reshaped
+
+
+def _trace(fn, inps) -> torch.fx.GraphModule:  # type: ignore[no-untyped-def]
+    fake_mode = maybe_get_fake_mode(inps[0][0])
+    assert fake_mode is not None
+    with fake_mode, enable_python_dispatcher():
+        return make_fx(fn)(*inps)
+
+
+def _insert_fn_trace_before_node(  # type: ignore[no-untyped-def]
+    g: torch.fx.Graph,
+    fn_to_trace,
+    inps,
+    insert_before_node,
+    g_fn_inps,
+    g_fn_outs,
+) -> dict[torch.fx.Node, torch.fx.Node]:  # type: ignore[no-untyped-def]
+    fn_gm = _trace(
+        fn_to_trace,
+        inps,
+    )
+    fn_g = fn_gm.graph
+    fn_g_ins = fn_g.find_nodes(op="placeholder")
+    env = {fn_g_ins[idx]: g_fn_inps[idx] for idx in range(len(g_fn_inps))}
+    g_fn_new_outs: list[torch.fx.Node] = []
+    with g.inserting_before(insert_before_node):
+        for _n in fn_g.nodes:
+            if _n.op == "placeholder":
+                continue
+            _new_n = g.node_copy(_n, lambda x: env[x])
+            env[_n] = _new_n
+            if _n.op == "output":
+                g_fn_new_outs = _new_n.args[0]  # type: ignore[assignment]
+                g.erase_node(_new_n)
+    replacements = {  # noqa: C416
+        orig_out: new_out for orig_out, new_out in zip(g_fn_outs, g_fn_new_outs)
+    }
+    for orig_out, new_out in zip(g_fn_outs, g_fn_new_outs):
+        orig_out.replace_all_uses_with(new_out)
+    return replacements
+
+
 def merge_reduce_scatter(
     gm: torch.fx.GraphModule, rs_buckets: list[list[torch.fx.Node]]
 ) -> None:
+    trace_structured(
+        "artifact",
+        metadata_fn=lambda: {
+            "name": "fx_bucketing_passes_reduce_scatter_buckets",
+            "encoding": "string",
+        },
+        payload_fn=lambda: str(rs_buckets),
+    )
+    n_buckets = len(rs_buckets)
+    g = gm.graph
+    rs_ins: list[list[torch.fx.Node]] = [[] for _ in range(n_buckets)]
+    rs_waits: list[list[torch.fx.Node]] = [[] for _ in range(n_buckets)]
+    rs_to_bucket_idx: dict[torch.fx.Node, int] = {}
+    for bucket_idx, rs_nodes in enumerate(rs_buckets):
+        for rs_node in rs_nodes:
+            rs_to_bucket_idx[rs_node] = bucket_idx
+
+    for bucket_idx, rs_bucket in enumerate(rs_buckets):
```

Comment:
nit: repeating the same for loop while naming the variable differently. maybe just merge the loops?